rfcs/0008-transition-frontier-controller.md

```diff
@@ -0,0 +1,337 @@
+changes in behavior and structure in more detail in [big Changes](#big-changes)
+
+<a href="big-changes"></a>
+### Big Changes
+
+Major differences in behavior between this component and the existing ledger-
+builder-controller are as follows:
+
+1. We only sync-ledger to the locked-ledger (if necessary) and do a new
+[History sync](#history-sync) for catchup jobs
+
+Rationale: Without history sync, we are vulnerable to believing an invalid
+state: Certain parts of the ledger-builder state are not checked in a SNARK and
+so we must get enough info to go back to the locked state (where we know we've
+achieved consensus).
+
+2. We take advantage of the merkle-mask to instead of transitions, putting a `Breadcrumb.t` which contains an `External_transition.t` and a now light-weight `Ledger_builder.t` it also has a notion of the prior breadcrumb. See [RFC-0007](0007-persistent-ledger-builder-controller) for more info on merkle masks, and [Transition frontier](#transition-frontier) for more info about how these masks are configured.
+
+Rationale: We no longer need the asynchronous `Path_traversal` job that is present in the current ledger-builder-controller (simplifying complexity), and the frequent operation of answering sync-ledger queries no longer require materializing ledger-builders.
+
+3. The `Ktree.t` is mutable and exposes an $O(1)$ `lookup : State_hash.t -> node_entry` where `node_entry` contains (in our case) a `Breadcrumb.t`. We will discuss this further in [Ktree](#ktree).
+
+Rationale: We perform `lookup` and `add` frequently on this structure and `add` can be $O(1)$ in the presence of this `lookup` too. `lookup` also lets us traverse the tree forwards and backwards without explicit backedges in our [rose tree](https://en.wikipedia.org/wiki/Rose_tree) -- further simplifying the `Ktree.t` implementation.
+
+<a href="transition-handler"></a>
+### Transition Handler
+
+The transition handler is broken down into two main components: Validator and
+Processor
+
+#### Validator
+
+* Input: `External_transition.t`
+* Output: `External_transition.t` (we should consider making an `External_transition.Validated.t` for output here)
+
+The validator receives as input external transitions from the network and
+performs a series of checks on the transition before passing it off to be
+processed. For simplicity it can also receive transitions created by the local
+proposer. In the future, we can consider skipping this step.
+
+Particularly, we validate transitions with the following checks (in this order):
+
+1. Checking for existence in the breadcrumb tree (we've already seen it)
+2. Checking for consensus errors
+3. Verifying the included SNARK
+
+This order is chosen because it's cheaper computationally to perform (1) and
+(2). So if we get lucky we can short-circuit before getting to the SNARK step.
+
+#### Processor
+
+* Input: `External_transition.t` (from validator) and `Breadcrumb.t` (from catchup)
```

Comment:
Actually a `Breadcrumb.t list` for all of the breadcrumbs that were returned during a catchup.

Reply:
Also, should discuss priorities between these two input pipes. I think the catchup pipe should probably have priority.

Reply:
I think it can't be a breadcrumb list (see my note on your PR). We should discuss this.